be/src/runtime/tablets_channel.cpp

```diff
@@ -49,115 +58,168 @@ TabletsChannel::~TabletsChannel() {
     _row_desc = new RowDescriptor(_tuple_desc, false);
 
-    _num_remaining_senders = params.num_senders();
-    _next_seqs.resize(_num_remaining_senders, 0);
-    _closed_senders.Reset(_num_remaining_senders);
+    _num_remaining_senders.store(params.num_senders(), std::memory_order_release);
+    _senders = std::vector<Sender>(params.num_senders());
 
     RETURN_IF_ERROR(_open_all_writers(params));
-
-    _state = kOpened;
     return Status::OK();
 }
 
-Status TabletsChannel::add_chunk(const PTabletWriterAddChunkRequest& params) {
-    {
-        std::lock_guard<std::mutex> l(_global_lock);
-        if (_state == kFinished) {
-            return _close_status;
-        }
+void TabletsChannel::add_chunk(brpc::Controller* cntl, const PTabletWriterAddChunkRequest& request,
+                               PTabletWriterAddBatchResult* response, google::protobuf::Closure* done) {
+    ClosureGuard done_guard(done);
 
-        auto next_seq = _next_seqs[params.sender_id()];
-        // check packet
-        if (params.packet_seq() < next_seq) {
-            LOG(INFO) << "packet has already received before, expect_seq=" << next_seq
-                      << ", received_seq=" << params.packet_seq();
-            return Status::OK();
-        } else if (params.packet_seq() > next_seq) {
-            LOG(WARNING) << "lost data packet, expect_seq=" << next_seq << ", received_seq=" << params.packet_seq();
-            return Status::InternalError("lost data packet");
-        }
+    auto t0 = std::chrono::steady_clock::now();
+
+    if (UNLIKELY(!request.has_sender_id())) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("no sender_id in PTabletWriterAddChunkRequest");
+        return;
+    }
+    if (UNLIKELY(request.sender_id() < 0)) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("negative sender_id in PTabletWriterAddChunkRequest");
+        return;
+    }
+    if (UNLIKELY(request.sender_id() >= _senders.size())) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs(
+                fmt::format("invalid sender_id {} in PTabletWriterAddChunkRequest, limit={}", request.sender_id(),
+                            _senders.size()));
+        return;
+    }
+    if (UNLIKELY(!request.has_packet_seq())) {
+        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
+        response->mutable_status()->add_error_msgs("no packet_seq in PTabletWriterAddChunkRequest");
+        return;
     }
 
-    auto& pchunk = params.chunk();
-    {
-        std::lock_guard<std::mutex> l(_global_lock);
-        if (_chunk_meta.types.empty()) {
-            RETURN_IF_ERROR(_build_chunk_meta(pchunk));
-        }
+    Sender& sender = _senders[request.sender_id()];
```

Comment:
const Sender& sender

Reply:
The sender will modified later, see the next_id